crypto/rsa/rsa_sp800_56b_gen.c

```diff
@@ -0,0 +1,365 @@
+                                  BIGNUM *Xpout, const BIGNUM *Xp,
+                                  const BIGNUM *Xp1, const BIGNUM *Xp2,
+                                  BIGNUM *q1, BIGNUM *q2, BIGNUM *Xqout,
+                                  const BIGNUM *Xq, const BIGNUM *Xq1,
+                                  const BIGNUM *Xq2, int nbits,
+                                  const BIGNUM *e, BN_CTX *ctx, BN_GENCB *cb)
+{
+    int ret = 0, ok;
+    BIGNUM *Xpo = NULL, *Xqo = NULL, *tmp = NULL;
+
+    /* (Step 1) Check key length
+     * NOTE: SP800-131A Rev1 Disallows key lengths of < 2048 bits for RSA
+     * Signature Generation and Key Agree/Transport.
+     */
+    if (nbits != 2048 && nbits != 3072) {
+        RSAerr(RSA_F_RSA_FIPS186_4_GEN_PROB_PRIMES, RSA_R_INVALID_KEY_LENGTH);
+        return 0;
+    }
+
+    if (!rsa_check_public_exponent(e)) {
+        RSAerr(RSA_F_RSA_FIPS186_4_GEN_PROB_PRIMES,
+               RSA_R_PUB_EXPONENT_OUT_OF_RANGE);
+        goto err;
+    }
+
+#if 0
+    /* (Step 3) Determine strength and check rand generator strength is ok */
+    if (!rsa_strength_check_prng(nbits))
+        goto err;
+#endif
+
+    BN_CTX_start(ctx);
+    tmp = BN_CTX_get(ctx);
+    Xpo = (Xpout != NULL) ? Xpout : BN_CTX_get(ctx);
+    Xqo = (Xqout != NULL) ? Xqout : BN_CTX_get(ctx);
+    if (tmp == NULL || Xpo == NULL || Xqo == NULL)
+        goto err;
+
+    if (rsa->p == NULL)
+        rsa->p = BN_secure_new();
+    if (rsa->q == NULL)
+        rsa->q = BN_secure_new();
+    if (rsa->p == NULL || rsa->q == NULL)
+        goto err;
+
+    for (;;) {
+        /* (Step 4) Generate p, Xp */
+        if (!bn_rsa_fips186_4_gen_prob_primes(rsa->p, Xpo, p1, p2, Xp, Xp1, Xp2,
+                                              nbits, e, ctx, cb))
+            goto err;
+        for(;;) {
+            /* (Step 5) Generate q, Xq*/
+            if (!bn_rsa_fips186_4_gen_prob_primes(rsa->q, Xqo, q1, q2, Xq, Xq1,
+                                                  Xq2, nbits, e, ctx, cb))
+                goto err;
+
+            /* (Step 6) |Xp - Xq| > 2^(nbitlen/2 - 100) */
+            ok = rsa_check_pminusq_diff(tmp, Xpo, Xqo, nbits);
+            if (ok < 0)
+                goto err;
+            if (ok == 0)
+                continue;
+
+            /* (Step 6) |p - q| > 2^(nbitlen/2 - 100) */
+            ok = rsa_check_pminusq_diff(tmp, rsa->p, rsa->q, nbits);
+            if (ok < 0)
+                goto err;
+            if (ok == 0)
+                continue;
+            goto end; /* successfully finished */
+        }
+    }
+end:;
+    ret = 1;
+err:
+    /* Zeroize any internally generated values that are not returned */
+    if (Xpo != Xpout)
+        BN_clear(Xpo);
+    if (Xqo != Xqout)
+        BN_clear(Xqo);
+    BN_clear(tmp);
+
+    BN_CTX_end(ctx);
+    return ret;
+}
+
+#if 0
+/* TODO - How do we implement strength for fips??  - it should be seperated
+ * from the RNG
```

Comment:
If you call RAND_DRBG_get0_private(), and then check drbg->strength you should get the value.

Reply:
That would give the strength of the DRBG (which is probably set to 256).
In the current FIPS specifications security strength is associated with keys.
i.e-
(1) to protect data (encryption, signing, wrapping, key gen) a minimum strength of 112 is required.
(2) to process data (decryption, verifying, unwrapping) a minimum strength of 80 is required.
Note that it is smaller because of support for deprecated key lengths.
-------
I am thinking that security strength is more of a global concept, which is not just related to DRBG's.
When generating any keys it would use the minimum protect strength, so the DRBG should only really need to generate this strength (not the maximum strength of the DRBG). 
-------

Reply:
removed and explained why it was removed